### ISSUE TITLE
WaveFlow vocoder

### ISSUE THREAD

**Issue by qo4on** [espnet/espnet#1798]
Recently Baidu [released](https://github.com/PaddlePaddle/Parakeet#pre-trained-models-and-audio-samples) mel2wav pretrained model which is close to WaveNet in terms of quality but much faster. Its config:

```
valid_size: 16
segment_length: 16000
sample_rate: 22050
fft_window_shift: 256
fft_window_size: 1024
fft_size: 1024
mel_bands: 80
mel_fmin: 0.0
mel_fmax: 8000.0

seed: 1234
learning_rate: 0.0002
batch_size: 8
test_every: 2000
save_every: 10000
max_iterations: 3000000

sigma: 1.0
n_flows: 8
n_group: 16
n_layers: 8
n_channels: 64
kernel_h: 3
kernel_w: 3
```

I have a few questions:
1. Can this model work with your `phn_train_no_dev_pytorch_train_tacotron2.v3_fastspeech.v4.single`?
2. Are you going to train WaveFlow?

I [managed to run](https://github.com/PaddlePaddle/models/issues/4523) inference of the WaveFlow, but the output file turned out to be a noise.  This was my first try of PaddlePaddle ML framework, so maybe I missed something.

**Comment by kan-bayashi** [comment]
1. Please check the following items:
	- mel range (mel_fmin and mel_fmax)
	- normalization (we perform mean-var normalization. If the different normalization is performed in waveflow, you need to generate Mel -> de-normalization with epsnet -> normalization with WaveFlow -> decode with WaveFlow
)
2. Currently, no. Since I'm focusing on GAN-based vocoder (MelGAN and Parallel WaveGAN). Do you think WaveFlow is better than Parallel WaveGAN or MelGAN?

**Comment by qo4on** [comment]
Thanks for your answer.
1. I opened a test audio (`phn_train_no_dev_pytorch_train_tacotron2.v3_fastspeech.v4.single` and `Parallel WaveGAN` from your Colab example) in Audition and found much wider 0-11000 Hz mel range than 80-7600 Hz:
![image](https://user-images.githubusercontent.com/35609308/79114647-3cb42d80-7d9d-11ea-8f26-48c3e730a8d1.png)
Can you clarify how these frequencies are related? Why did you switch from `0-8000` to `80-7600`?

2. [Baidu claims](https://arxiv.org/pdf/1912.01219.pdf) that compare to WaveNet they reached better quality in terms of log-likelihoods and the same in MOS. But WaveNet is practically not applicable due to very slow inference. On the other hand WaveFlow is much faster.
I didn't managed to test WaveFlow because their github code generates noise. So, I can't tell for sure. But I like the examples they released. Parallel WaveGAN does a great job, but for me it has some metallic artifacts. MelGAN quality is lower than Parallel WaveGAN.

**Comment by kan-bayashi** [comment]
> Can you clarify how these frequencies are related?
> Why did you switch from 0-8000 to 80-7600?

I have some discussions (see https://github.com/espnet/espnet/issues/1192).
According to preliminary experiments, the limit of the Mel range can alleviate the over-smoothing problem.

2. OK. I will check their samples. Thank you for your info.

**Comment by qo4on** [comment]
I found that you have the same wrong values in the `config.yml` for `pwg_v3`.
Can you tell the difference between these two models, am I right that `v3` is better?

```
train_nodev_ljspeech_parallel_wavegan.v3/checkpoint-3000000steps.pkl
ljspeech.parallel_wavegan.v2/checkpoint-400000steps.pkl
```

**Comment by kan-bayashi** [comment]
Note that the mel range is different from the range of frequency in the waveform.
Even if the mel range is limited (e.g. 80-7600 Hz) but the actual waveform has full range frequency components (0-11025 Hz).

I will show the table of the relationship between [espnet pretrained model](https://github.com/espnet/espnet#tts-results) and [my PWG pretrained model links](https://github.com/kan-bayashi/ParallelWaveGAN#results).
| ESPnet | PWG | 
| :----: | :----: |
| v1  | v1.nolimit |
| v2 | v1 |
| -  | v3|

Sorry for confusion.
Basically, v1 and v2 in ESPnet's table is the same except for the Mel range.
v3 uses Parallel WaveGAN Generator + MelGAN multi-scale discriminator.
See the detail configuration in the link to the confing file (https://github.com/kan-bayashi/ParallelWaveGAN#results).
I've not yet compared these quality with subjective evaluation.

**Comment by qo4on** [comment]
Thank you. Your wavegan.v3 generates less background noise than wavegan.v2. But overall quality seems to be a bit better in wavegan.v2.

I managed to run inference of WaveFlow. It's impressive! I don't hear the difference between generated file and the original. Of course it is for ground truth mel from original `wav`, not from generated mel.

Can you tell how to do `de-normalization with espnet's statistics` for your fastspeach model `phn_train_no_dev_pytorch_train_tacotron2.v3_fastspeech.v4.single`?

If you're interested you can run WaveFlow inference with this code, just copypaste it to Colab:
```
!sudo apt-get update -y -qqq --fix-missing && apt-get install -y -qqq libsndfile1 > /dev/null
!pip install -U -qqq imgaug scipy albumentations paddlepaddle-gpu librosa > /dev/null
!git clone -qqq https://github.com/PaddlePaddle/Parakeet > /dev/null
%cd /content/Parakeet
!pip install -qqq -e . > /dev/null

import os, pathlib, nltk, sys
import numpy as np

nltk.download("punkt")
nltk.download("cmudict")

pth_1 = "/content/Parakeet"
if pth_1 not in sys.path: sys.path.insert(0, pth_1)

folder = '/content/downloads/'
os.makedirs(folder, exist_ok=True)
%cd {folder}
if not os.path.exists(folder + 'LJ001-0001.wav'):
  !curl -Ls https://dl.dropboxusercontent.com/s/jj78665lrhiod97/LJ001-0001.wav -o LJ001-0001.wav
wav_path = os.path.join(folder, 'LJ001-0001.wav')
if not os.path.exists('/content/downloads/waveflow_res128_ljspeech_ckpt_1.0.zip'):
  !wget -qqq https://paddlespeech.bj.bcebos.com/Parakeet/waveflow_res128_ljspeech_ckpt_1.0.zip > /dev/null
  !unzip -qqq /content/downloads/waveflow_res128_ljspeech_ckpt_1.0.zip -d /content/Parakeet/examples/waveflow > /dev/null
  !rm -rf /content/sample_data /content/downloads/waveflow_res128_ljspeech_ckpt_1.0.zip

!export CUDA_VISIBLE_DEVICES=0
from parakeet.models.waveflow import WaveFlowModule
from parakeet.modules import weight_norm
from parakeet.utils import io
import paddle.fluid.dygraph as dg
from paddle import fluid
from argparse import Namespace
from scipy.io.wavfile import read
from scipy.io.wavfile import write
from pprint import pprint
from ruamel import yaml
import random, librosa

class WaveFlow():
  def __init__(self,
               config,
               parallel=False,
               rank=0,
               nranks=1,
               tb_logger=None):
    self.config = config
    self.parallel = parallel
    self.rank = rank
    self.nranks = nranks
    self.tb_logger = tb_logger
    self.dtype = "float16" if config.use_fp16 else "float32"

  def build(self):
    config = self.config
    waveflow = WaveFlowModule(config)
    iteration = io.load_parameters(
        model=waveflow,
        checkpoint_path=config.checkpoint)
    
    for layer in waveflow.sublayers():
      if isinstance(layer, weight_norm.WeightNormWrapper):
        layer.remove_weight_norm()    
    self.waveflow = waveflow

    return iteration

  @dg.no_grad
  def infer(self, mel):
    self.waveflow.eval()
    config = self.config
    
    audio = self.waveflow.synthesize(mel, sigma=self.config.sigma)
    audio = audio[0]

    # Denormalize audio from [-1, 1] to [-32768, 32768] int16 range
    audio = audio.numpy().astype("float32") * 32768.0
    audio = audio.astype('int16')
    write(os.path.join(config.output, 'test.wav'), config.sample_rate, audio)

def get_mel(audio, config):
  spectrogram = librosa.core.stft(
      audio,
      n_fft=config.fft_size,
      hop_length=config.fft_window_shift,
      win_length=config.fft_window_size)
  spectrogram_magnitude = np.abs(spectrogram)

  # mel_filter_bank shape: [n_mels, 1 + n_fft/2]
  mel_filter_bank = librosa.filters.mel(sr=config.sample_rate,
                                        n_fft=config.fft_size,
                                        n_mels=config.mel_bands,
                                        fmin=config.mel_fmin,
                                        fmax=config.mel_fmax)
  # mel shape: [n_mels, num_frames]
  mel = np.dot(mel_filter_bank, spectrogram_magnitude)

  # Normalize mel
  clip_val = 1e-5
  ref_constant = 1
  mel = np.log(np.clip(mel, a_min=clip_val, a_max=None) * ref_constant)

  return mel

config_path = '/content/Parakeet/examples/waveflow/waveflow_res128_ljspeech_ckpt_1.0/waveflow_ljspeech.yaml'
with open(config_path) as f:
  config = yaml.load(f, Loader=yaml.Loader)

config = Namespace(**config,
                   checkpoint='/content/Parakeet/examples/waveflow/waveflow_res128_ljspeech_ckpt_1.0/step-2000000',
                   config=config_path,
                   output='/content',
                   use_fp16=True,
                   use_gpu=True
                   )

pprint(vars(config))

place = fluid.CUDAPlace(0) if config.use_gpu else fluid.CPUPlace()
with dg.guard(place):
  # Fix random seed
  seed = config.seed
  random.seed(seed)
  np.random.seed(seed)
  fluid.default_startup_program().random_seed = seed
  fluid.default_main_program().random_seed = seed

  # Build model
  model = WaveFlow(config)
  iteration = model.build()

  # Load audio
  loaded_sr, audio = read(wav_path)

  # Normalize audio to the [-1, 1] range
  audio = audio.astype(np.float32) / 32768.0
  
  mel = dg.to_variable(np.expand_dims(get_mel(audio, config), axis=0))
  print(mel.numpy().shape, np.min(mel.numpy()), np.max(mel.numpy()))
  model.infer(mel)

{'batch_size': 8,
 'checkpoint': '/content/Parakeet/examples/waveflow/waveflow_res128_ljspeech_ckpt_1.0/step-2000000',
 'config': '/content/Parakeet/examples/waveflow/waveflow_res128_ljspeech_ckpt_1.0/waveflow_ljspeech.yaml',
 'fft_size': 1024,
 'fft_window_shift': 256,
 'fft_window_size': 1024,
 'kernel_h': 3,
 'kernel_w': 3,
 'learning_rate': 0.0002,
 'max_iterations': 3000000,
 'mel_bands': 80,
 'mel_fmax': 8000.0,
 'mel_fmin': 0.0,
 'n_channels': 128,
 'n_flows': 8,
 'n_group': 16,
 'n_layers': 8,
 'output': '/content',
 'sample_rate': 22050,
 'save_every': 10000,
 'seed': 1234,
 'segment_length': 16000,
 'sigma': 1.0,
 'test_every': 2000,
 'use_fp16': True,
 'use_gpu': True,
 'valid_size': 16}
/usr/local/lib/python3.6/dist-packages/paddle/fluid/data_feeder.py:93: UserWarning: The data type of 'input' in assign only support float16 in GPU now. (When the type of input in assign is Variable.)
  (input_name, op_name, extra_message))
/usr/local/lib/python3.6/dist-packages/paddle/fluid/data_feeder.py:93: UserWarning: The data type of 'x' in cast only support float16 in GPU now. 
  (input_name, op_name, extra_message))
[checkpoint] Rank 0: loaded model from /content/Parakeet/examples/waveflow/waveflow_res128_ljspeech_ckpt_1.0/step-2000000.pdparams
(1, 80, 832) -11.512925 1.4659
/usr/local/lib/python3.6/dist-packages/paddle/fluid/data_feeder.py:93: UserWarning: The data type of 'input' in squeeze only support float16 in GPU now. 
  (input_name, op_name, extra_message))
```

**Comment by kan-bayashi** [comment]
Sorry for the late reply.[
Thanks! I will try!

> Can you tell how to do de-normalization with espnet's statistics for your fastspeech model phn_train_no_dev_pytorch_train_tacotron2.v3_fastspeech.v4.single?

You can use `cmvn.ark` file as follows:
```
import kaldiio
from espnet.transform.cmvn import CMVN
cmvn_path = "path_to_cmvn_file"
stats = kaldiio.load_mat(cmvn_path)
cmvn = CMVN(stats, norm_means=True, norm_vars=True, reverse=True)
feature_denorm = cmvn(features)  # (#frames, #dim) 
```

**Comment by qo4on** [comment]
It's not working for some reason:
```
import kaldiio
from espnet.transform.cmvn import CMVN
cmvn_path = "/content/downloads/en/fastspeech_v4_t2_v3_phn/data/phn_train_no_dev/cmvn.ark"
stats = kaldiio.load_mat(cmvn_path)
cmvn = CMVN(stats, norm_means=True, norm_vars=True, reverse=True)

AttributeError                            Traceback (most recent call last)
<ipython-input-18-8818b83a6ab9> in <module>()
      3 cmvn_path = "/content/downloads/en/fastspeech_v4_t2_v3_phn/data/phn_train_no_dev/cmvn.ark"
      4 stats = kaldiio.load_mat(cmvn_path)
----> 5 cmvn = CMVN(stats, norm_means=True, norm_vars=True, reverse=True)

2 frames
/content/espnet/espnet/transform/cmvn.py in __init__(self, stats, norm_means, norm_vars, filetype, utt2spk, spk2utt, reverse, std_floor)
     20             # Use for global CMVN
     21             if filetype == 'mat':
---> 22                 stats_dict = {None: kaldiio.load_mat(stats)}
     23             # Use for global CMVN
     24             elif filetype == 'npy':

/usr/local/lib/python3.6/dist-packages/kaldiio/matio.py in load_mat(ark_name, endian, as_bytes)
    199 def load_mat(ark_name, endian='<', as_bytes=False):
    200     assert endian in ('<', '>'), endian
--> 201     ark, offset, slices = _parse_arkpath(ark_name)
    202     with open_like_kaldi(ark, 'rb') as fd:
    203         return _load_mat(fd, offset, slices, endian=endian, as_bytes=as_bytes)

/usr/local/lib/python3.6/dist-packages/kaldiio/matio.py in _parse_arkpath(ark_name)
    223     """
    224 
--> 225     if ark_name.rstrip()[-1] == '|' or ark_name.rstrip()[0] == '|':
    226         # Something like: "| cat foo" or "cat bar|" shouldn't be parsed
    227         return ark_name, None, None

AttributeError: 'numpy.ndarray' object has no attribute 'rstrip'
```

**Comment by kan-bayashi** [comment]
Sorry, I was wrong. I updated code.

**Comment by qo4on** [comment]
Thank you. I did everything from your list:

> * generate Mel
> * de-normalization with espnet's statistics
> * normalization with WaveFlow's one
> * decode with WaveFlow.

I keep only the normalization part of WaveFlow's `get_mel` function. But the `wav` file I get is not a sound I expected. It's just some quiet sound artifacts.

Can you point out what I did wrong?

```
class WaveFlow():
  def __init__(self,
              config,
              parallel=False,
              rank=0,
              nranks=1,
              tb_logger=None):
    self.config = config
    self.parallel = parallel
    self.rank = rank
    self.nranks = nranks
    self.tb_logger = tb_logger
    self.dtype = "float16" if config.use_fp16 else "float32"

  def build(self):
    config = self.config
    waveflow = WaveFlowModule(config)
    iteration = io.load_parameters(
        model=waveflow,
        checkpoint_path=config.checkpoint)
    
    for layer in waveflow.sublayers():
      if isinstance(layer, weight_norm.WeightNormWrapper):
        layer.remove_weight_norm()    
    self.waveflow = waveflow

    return iteration

  @dg.no_grad
  def infer(self, mel):
    self.waveflow.eval()
    config = self.config
    
    audio = self.waveflow.synthesize(mel, sigma=self.config.sigma)
    audio = audio[0]

    # Denormalize audio from [-1, 1] to [-32768, 32768] int16 range
    audio = audio.numpy().astype("float32") * 32768.0
    audio = audio.astype('int16')
    write(os.path.join(config.output, 'test.wav'), config.sample_rate, audio)

def norm_mel(mel):
  # Normalize mel
  clip_val = 1e-5
  ref_constant = 1
  mel = np.log(np.clip(mel, a_min=clip_val, a_max=None) * ref_constant)

  return mel

config_path = '/content/Parakeet/examples/waveflow/waveflow_res128_ljspeech_ckpt_1.0/waveflow_ljspeech.yaml'
with open(config_path) as f:
  config = yaml.load(f, Loader=yaml.Loader)

config = Namespace(**config,
                  checkpoint='/content/Parakeet/examples/waveflow/waveflow_res128_ljspeech_ckpt_1.0/step-2000000',
                  config=config_path,
                  output='/content',
                  use_fp16=True,
                  use_gpu=True
                  )

pprint(vars(config))

def get_audio():
  with torch.no_grad():
    x = frontend(input_text)
    c, _, _ = t2m_model.inference(x, inference_args)  # Output sequence of features (L, odim)
    print(c.shape, torch.min(c), torch.max(c))
    feature_denorm = cmvn(c.cpu()).numpy().T  # (#frames, #dim)
    print(feature_denorm.shape, np.min(feature_denorm), np.max(feature_denorm))

  !export CUDA_VISIBLE_DEVICES=0
  place = fluid.CUDAPlace(0) if config.use_gpu else fluid.CPUPlace()
  with dg.guard(place):
    # Fix random seed
    seed = config.seed
    random.seed(seed)
    np.random.seed(seed)
    fluid.default_startup_program().random_seed = seed
    fluid.default_main_program().random_seed = seed

    # Build model
    model = WaveFlow(config)
    iteration = model.build()

    mel = dg.to_variable(np.expand_dims(norm_mel(feature_denorm), axis=0))
    print(mel.numpy().shape, np.min(mel.numpy()), np.max(mel.numpy()))
    model.infer(mel)

# display(Audio(audio, rate=config.sample_rate))

get_audio()

{'batch_size': 8,
 'checkpoint': '/content/Parakeet/examples/waveflow/waveflow_res128_ljspeech_ckpt_1.0/step-2000000',
 'config': '/content/Parakeet/examples/waveflow/waveflow_res128_ljspeech_ckpt_1.0/waveflow_ljspeech.yaml',
 'fft_size': 1024,
 'fft_window_shift': 256,
 'fft_window_size': 1024,
 'kernel_h': 3,
 'kernel_w': 3,
 'learning_rate': 0.0002,
 'max_iterations': 3000000,
 'mel_bands': 80,
 'mel_fmax': 8000.0,
 'mel_fmin': 0.0,
 'n_channels': 128,
 'n_flows': 8,
 'n_group': 16,
 'n_layers': 8,
 'output': '/content',
 'sample_rate': 22050,
 'save_every': 10000,
 'seed': 1234,
 'segment_length': 16000,
 'sigma': 1.0,
 'test_every': 2000,
 'use_fp16': True,
 'use_gpu': True,
 'valid_size': 16}
torch.Size([509, 80]) tensor(-1.8778, device='cuda:0') tensor(2.8275, device='cuda:0')
(80, 509) -4.448508 0.6760279
/usr/local/lib/python3.6/dist-packages/paddle/fluid/data_feeder.py:93: UserWarning: The data type of 'input' in assign only support float16 in GPU now. (When the type of input in assign is Variable.)
  (input_name, op_name, extra_message))
/usr/local/lib/python3.6/dist-packages/paddle/fluid/data_feeder.py:93: UserWarning: The data type of 'x' in cast only support float16 in GPU now. 
  (input_name, op_name, extra_message))
[checkpoint] Rank 0: loaded model from /content/Parakeet/examples/waveflow/waveflow_res128_ljspeech_ckpt_1.0/step-2000000.pdparams
(1, 80, 509) -11.512925 -0.39152095
/usr/local/lib/python3.6/dist-packages/paddle/fluid/data_feeder.py:93: UserWarning: The data type of 'input' in squeeze only support float16 in GPU now. 
  (input_name, op_name, extra_message))
```

**Comment by kan-bayashi** [comment]
Sorry for the late reply.
As I mentioned before https://github.com/espnet/espnet/issues/1798#issuecomment-612822620, you need to match the mel range. And it is better to check the procedure of the extraction of mel-spectrogram is the same or not.

**Comment by qo4on** [comment]
Ok, thanks.
